File: src/two_stage.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
MIT License

Copyright (c) 2024 cauchy1988

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""


import copy
import random
from typing import Optional

from solution import PDWTWSolution
from alns import adaptive_large_neighbourhood_search


class TwoStageError(Exception):
    """Custom exception for two-stage algorithm operations"""
    pass


def first_stage_to_limit_vehicle_num_in_homogeneous_fleet(one_solution: PDWTWSolution) -> PDWTWSolution:
    """
    First stage: minimize the number of vehicles in a homogeneous fleet.
    
    This stage works in two phases:
    1. Insert all requests by adding vehicles as needed
    2. Iteratively remove vehicles and try to reassign requests
    
    Args:
        one_solution: Initial solution to optimize
        
    Returns:
        Optimized solution with minimal vehicle count
        
    Raises:
        TwoStageError: If algorithm fails to converge or encounters errors
    """
    if one_solution is None:
        raise ValueError("one_solution cannot be None")
    
    # Phase 1: Insert all requests by adding vehicles as needed
    requests_in_bank = one_solution.request_bank.copy()
    
    a_iteration_num = 0
    max_iterations = 1000  # Prevent infinite loops
    new_vehicle_add_flag = False
    
    while requests_in_bank and a_iteration_num < max_iterations:
        a_iteration_num += 1
        
        current_request = requests_in_bank.pop(0)
        if one_solution.insert_one_request_to_any_vehicle_route_optimal(current_request):
            new_vehicle_add_flag = False
            continue
        else:
            # Check if we're stuck in a loop
            if new_vehicle_add_flag:
                raise TwoStageError(f"Failed to insert request {current_request} even after adding new vehicle. "
                                  f"Algorithm may be stuck.")
            
            one_solution.add_one_same_vehicle()
            requests_in_bank.append(current_request)
            new_vehicle_add_flag = True
    
    if a_iteration_num >= max_iterations:
        raise TwoStageError(f"First stage failed to converge after {max_iterations} iterations")
    
    result_solution = one_solution.copy()
    
    # Phase 2: Iteratively remove vehicles and try to reassign requests
    total_iteration_num = a_iteration_num
    max_total_iterations = one_solution.meta_obj.parameters.theta
    
    while total_iteration_num <= max_total_iterations:
        # Get the vehicle with maximum ID to remove
        max_vehicle_id = one_solution.max_vehicle_id()
        if max_vehicle_id is None:
            break
        
        # Remove the vehicle and its route
        one_solution.delete_vehicle_and_its_route(max_vehicle_id)
        
        # Temporarily modify parameters for ALNS
        original_iteration_num = one_solution.meta_obj.parameters.iteration_num
        one_solution.meta_obj.parameters.iteration_num = one_solution.meta_obj.parameters.tau
        
        try:
            # Try to reassign requests using ALNS
            adaptive_large_neighbourhood_search(one_solution.meta_obj, one_solution, insert_unlimited=True)
            
            if not one_solution.request_bank:
                # Successfully reassigned all requests
                result_solution = one_solution.copy()
                total_iteration_num += one_solution.meta_obj.parameters.iteration_num
            else:
                # Failed to reassign all requests, stop here
                break
        finally:
            # Restore original parameters
            one_solution.meta_obj.parameters.iteration_num = original_iteration_num
    
    # Reset parameters to original values
    result_solution.meta_obj.parameters.reset()
    return result_solution


def two_stage_algorithm_in_homogeneous_fleet(initial_solution: PDWTWSolution) -> PDWTWSolution:
    """
    Two-stage algorithm for solving PDWTW with homogeneous fleet.
    
    Stage 1: Minimize the number of vehicles
    Stage 2: Optimize the solution using ALNS
    
    Args:
        initial_solution: Initial solution to optimize
        
    Returns:
        Optimized solution
        
    Raises:
        TwoStageError: If algorithm fails to converge
        ValueError: If input is invalid
    """
    if initial_solution is None:
        raise ValueError("initial_solution cannot be None")
    
    if not hasattr(initial_solution, 'meta_obj') or initial_solution.meta_obj is None:
        raise ValueError("initial_solution must have a valid meta_obj")
    
    try:
        # Stage 1: Minimize vehicle count
        result_solution = first_stage_to_limit_vehicle_num_in_homogeneous_fleet(initial_solution)
        
        # Stage 2: Optimize using ALNS
        adaptive_large_neighbourhood_search(result_solution.meta_obj, result_solution, insert_unlimited=True)
        
        return result_solution
        
    except Exception as e:
        raise TwoStageError(f"Two-stage algorithm failed: {str(e)}") from e


def _validate_solution(one_solution: PDWTWSolution) -> None:
    """
    Validate that a solution is ready for two-stage algorithm.
    
    Args:
        one_solution: Solution to validate
        
    Raises:
        ValueError: If solution is invalid
    """
    if one_solution is None:
        raise ValueError("Solution cannot be None")
    
    if not hasattr(one_solution, 'meta_obj'):
        raise ValueError("Solution must have meta_obj attribute")
    
    if not hasattr(one_solution, 'request_bank'):
        raise ValueError("Solution must have request_bank attribute")
    
    if not hasattr(one_solution, 'vehicle_bank'):
        raise ValueError("Solution must have vehicle_bank attribute")
    
    if not hasattr(one_solution.meta_obj, 'parameters'):
        raise ValueError("Solution meta_obj must have parameters attribute")


def safe_two_stage_algorithm(initial_solution: PDWTWSolution) -> PDWTWSolution:
    """
    Safely execute two-stage algorithm with input validation.
    
    Args:
        initial_solution: Initial solution to optimize
        
    Returns:
        Optimized solution
        
    Raises:
        TwoStageError: If algorithm fails
    """
    _validate_solution(initial_solution)
    return two_stage_algorithm_in_homogeneous_fleet(initial_solution)

File: java-version/src/main/java/com/tc/pdwtw/algorithm/TwoStage.java
package com.tc.pdwtw.algorithm;

import com.tc.pdwtw.model.*;
import java.util.*;

/**
 * MIT License
 * 
 * Copyright (c) 2024 cauchy1988
 * 
 * Two-stage algorithm for PDWTW with homogeneous fleet
 */
public class TwoStage {

    public static class TwoStageError extends RuntimeException {
        public TwoStageError(String message) {
            super(message);
        }
    }

    public static class AlnsResult {
        public final PDWTWSolution solution;
        public final int iterationCount;

        public AlnsResult(PDWTWSolution solution, int iterationCount) {
            this.solution = solution;
            this.iterationCount = iterationCount;
        }
    }

    /**
     * First stage: minimize the number of vehicles in a homogeneous fleet.
     * 
     * This stage works in two phases:
     * 1. Insert all requests by adding vehicles as needed
     * 2. Iteratively remove vehicles and try to reassign requests
     * 
     * @param oneSolution Initial solution to optimize
     * @return Optimized solution with minimal vehicle count
     * @throws TwoStageError If algorithm fails to converge or encounters errors
     */
    public static PDWTWSolution firstStageToLimitVehicleNumInHomogeneousFleet(PDWTWSolution oneSolution) {
        if (oneSolution == null) {
            throw new IllegalArgumentException("oneSolution cannot be null");
        }

        // Phase 1: Insert all requests by adding vehicles as needed
        List<Integer> requestsInBank = new ArrayList<>(oneSolution.getRequestBank());
        
        int aIterationNum = 0;
        int maxIterations = 1000; // Prevent infinite loops
        boolean newVehicleAddFlag = false;

        while (!requestsInBank.isEmpty() && aIterationNum < maxIterations) {
            aIterationNum++;

            Integer currentRequest = requestsInBank.remove(0);
            if (oneSolution.insertOneRequestToAnyVehicleRouteOptimal(currentRequest)) {
                newVehicleAddFlag = false;
                continue;
            } else {
                // Check if we're stuck in a loop
                if (newVehicleAddFlag) {
                    throw new TwoStageError("Failed to insert request " + currentRequest + 
                                          " even after adding new vehicle. Algorithm may be stuck.");
                }

                oneSolution.addOneSameVehicle(null);
                requestsInBank.add(currentRequest);
                newVehicleAddFlag = true;
            }
        }

        if (aIterationNum >= maxIterations) {
            throw new TwoStageError("First stage failed to converge after " + maxIterations + " iterations");
        }

        PDWTWSolution resultSolution = oneSolution.copyWithDeepCopiedMeta();
        
        System.out.println("vehicle num: " + oneSolution.getPaths().size());

        // Create a copy of vehicle_bank to avoid modification during iteration
        List<Integer> vehicleBankCopy = new ArrayList<>(oneSolution.getVehicleBank());
        for (Integer vehicleId : vehicleBankCopy) {
            oneSolution.deleteVehicleAndItsRoute(vehicleId);
        }

        // Phase 2: Iteratively remove vehicles and try to reassign requests
        int totalIterationNum = aIterationNum;
        int maxTotalIterations = oneSolution.getMetaObj().getParameters().getTheta();

        while (totalIterationNum <= maxTotalIterations) {
            // Get the vehicle with maximum ID to remove
            Integer maxVehicleId = oneSolution.maxVehicleId();
            if (maxVehicleId == null) {
                break;
            }

            System.out.println("loop num : " + totalIterationNum + ", vehicle num : " + oneSolution.getPaths().size());

            // Remove the vehicle and its route
            oneSolution.deleteVehicleAndItsRoute(maxVehicleId);

            // Temporarily modify parameters for ALNS
            int originalIterationNum = oneSolution.getMetaObj().getParameters().getIterationNum();
            oneSolution.getMetaObj().getParameters().setIterationNum(oneSolution.getMetaObj().getParameters().getTau());
            
            try {
                // Try to reassign requests using ALNS
                ALNS.AlnsResult alnsResult = ALNS.adaptiveLargeNeighbourhoodSearch(
                    oneSolution.getMetaObj(), oneSolution, true, true);
                
                if (alnsResult.bestSolution.getRequestBank().isEmpty()) {
                    // Successfully reassigned all requests
                    resultSolution = alnsResult.bestSolution.copyWithDeepCopiedMeta();
                } else {
                    // Failed to reassign all requests, stop here
                    break;
                }

                totalIterationNum += alnsResult.totalIterations;
            } catch (Exception e) {
                // Handle any errors during ALNS, including vehicle deletion errors
                System.out.println("Warning: ALNS failed during vehicle removal: " + e.getMessage());
                break;
            } finally {
                // Reset parameters to original values
                oneSolution.getMetaObj().getParameters().setIterationNum(originalIterationNum);
            }
        }

        // Reset parameters to original values
        resultSolution.getMetaObj().getParameters().reset();
        return resultSolution;
    }

    /**
     * Two-stage algorithm for solving PDWTW with homogeneous fleet.
     * 
     * Stage 1: Minimize the number of vehicles
     * Stage 2: Optimize the solution using ALNS
     * 
     * @param initialSolution Initial solution to optimize
     * @return Optimized solution
     * @throws TwoStageError If algorithm fails to converge
     */
    public static PDWTWSolution twoStageAlgorithmInHomogeneousFleet(PDWTWSolution initialSolution) {
        if (initialSolution == null) {
            throw new IllegalArgumentException("initialSolution cannot be null");
        }

        System.out.println("Starting two-stage algorithm");
        
        // Stage 1: Minimize vehicle count
        System.out.println("Stage 1: Minimizing vehicle count");
        PDWTWSolution stage1Solution = firstStageToLimitVehicleNumInHomogeneousFleet(initialSolution);
        
        System.out.println("Stage 1 completed. Vehicle count: " + stage1Solution.getPaths().size());
        
        // Stage 2: Optimize using ALNS
        System.out.println("Stage 2: Optimizing with ALNS");
        
        try {
            // Stage 2: Full ALNS optimization
            ALNS.AlnsResult finalResult = ALNS.adaptiveLargeNeighbourhoodSearch(
                stage1Solution.getMetaObj(), stage1Solution, false, false);
            System.out.println("Stage 2 completed. Final objective cost: " + finalResult.bestSolution.getObjectiveCost());
            return finalResult.bestSolution;
        } catch (Exception e) {
            System.out.println("Warning: Stage 2 optimization failed: " + e.getMessage());
            return stage1Solution;
        }
    }

}